pallets/capacity/src/lib.rs

```diff
@@ -0,0 +1,658 @@
+	capacity::{Nontransferable, Replenishable, TargetValidator},
+	msa::MessageSourceId,
+	utils::wrap_binary_data,
+};
+
+#[cfg(feature = "runtime-benchmarks")]
+use common_primitives::benchmarks::RegisterProviderBenchmarkHelper;
+
+pub use pallet::*;
+pub use types::*;
+pub use weights::*;
+pub mod types;
+
+#[cfg(test)]
+pub mod testing_utils;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[cfg(test)]
+mod extrinsics_tests;
+#[cfg(test)]
+mod helpers_tests;
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod types_tests;
+
+pub mod weights;
+
+type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+const STAKING_ID: LockIdentifier = *b"netstkng";
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	use frame_support::{pallet_prelude::*, Twox64Concat};
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::traits::{AtLeast32BitUnsigned, MaybeDisplay};
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// Function that allows a balance to be locked.
+		type Currency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+
+		/// Function that checks if an MSA is a valid target.
+		type TargetValidator: TargetValidator;
+
+		/// The minimum required token amount to stake. It facilitates cleaning dust when unstaking.
+		#[pallet::constant]
+		type MinimumStakingAmount: Get<BalanceOf<Self>>;
+
+		/// The maximum number of unlocking chunks a StakingAccountLedger can have. It determines how many concurrent unstaked chunks may exist.
+		#[pallet::constant]
+		type MaxUnlockingChunks: Get<u32>;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		/// A set of helper functions for benchmarking.
+		type BenchmarkHelper: RegisterProviderBenchmarkHelper;
+
+		/// The number of Epochs before you can unlock tokens after unstaking.
+		#[pallet::constant]
+		type UnstakingThawPeriod: Get<u16>;
+
+		/// Maximum number of blocks an epoch can be
+		/// currently used as the actual value of epoch length.
+		#[pallet::constant]
+		type MaxEpochLength: Get<Self::BlockNumber>;
+
+		/// A type that provides an Epoch number
+		/// traits pulled from frame_system::Config::BlockNumber
+		type EpochNumber: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ MaybeDisplay
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ sp_std::hash::Hash
+			+ MaxEncodedLen
+			+ TypeInfo;
+	}
+
+	/// Storage for keeping a ledger of staked token amounts for accounts.
+	/// - Keys: AccountId
+	/// - Value: [`StakingAccountDetails`](types::StakingAccountDetails)
+	#[pallet::storage]
+	#[pallet::getter(fn get_staking_account_for)]
+	pub type StakingAccountLedger<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, StakingAccountDetails<T>>;
```

Comment:
We should have done the same with Msa mappings but better late than never.

Reply:
Maybe I'm getting this wrong but aren't all keys generated from users wallet side?

Reply:
Yes. Here it's a bit different. To make this transaction, you must also have a private key associated with the account. That means that to make the tree unbalanced; someone must generate private keys associated with that account. It is difficult to generate private keys for arbitrary public keys.

This differs from creating an account in which you can send money to any random 32 bytes and create an account.

Reply:
How is it different form creating a new account? Private and public keys are generated together. So I can generate a lot of them and only use the ones that will cause imbalance in the tree. It would need a good number of token to create those accounts but doesn't mean that we will be immune to that problem.

Reply:
You would need to mine for specific private given public keys that would make it imbalanced. Trying to find the private key from the public key is really hard.

Reply:
@aramikm The difference is that you can "create an account" just by sending it tokens. You don't need the private key to "create an account". Here you must have a private key to trigger the creation of the "account"

Reply:
>This differs from creating an account in which you can send money to any random 32 bytes and create an account.

This means I can easily create accounts that would make the trees unbalanced. I can quickly generate those. There is no validation that address I am sending money has a private key associated with it. That means that I can be malicious by manipulating the storage tree.

In this case, in order to stake to an account a verification is done with the public key and data. So if I wanted to be malicious, I would have to find the public keys that I want that would make the tree unbalanced. Additionally, I would need the private keys to this. This is really hard.